File: src/ate_cloud/schemas/knowledge.py
"""Pydantic request/response schemas for the knowledge domain (task 10).

Boundary parse layer for the models in ``models/knowledge.py``. The CRUD
routers themselves are built later (task 12 ingestion, task 13 FMEA API,
task 15 diagnosis persistence); these schemas only define the wire shape:

- FMEA ratings are constrained to 1-10 at the boundary and ``rpn`` is NOT an
  accepted input — it is always derived server-side.
- Response schemas add id/timestamps and the computed ``rpn``.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

#: FMEA rating bounds — mirrors models/knowledge.py RATING_MIN/MAX.
RATING_MIN = 1
RATING_MAX = 10

RequirementSource = Literal["dsl", "atml", "manual"]


# ── TestRequirement ─────────────────────────────────────────────────────────


class TestRequirementCreate(BaseModel):
    """Create payload for a test requirement."""

    product_code: str = Field(..., min_length=1, max_length=100)
    requirement_code: str = Field(..., min_length=1, max_length=100)
    title: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    source: RequirementSource = "manual"
    atml_ref: str | None = Field(None, max_length=255)


class TestRequirementResponse(TestRequirementCreate):
    """A persisted test requirement."""

    model_config = {"from_attributes": True}

    id: str
    created_at: datetime
    updated_at: datetime


# ── TestCase ────────────────────────────────────────────────────────────────


class TestCaseCreate(BaseModel):
    """Create payload for a test case.

    ``requirement_id`` is nullable so cases can be ingested before their
    requirement (task 12 ingestion ordering) and linked later.
    """

    requirement_id: str | None = Field(None, max_length=36)
    case_code: str = Field(..., min_length=1, max_length=100)
    title: str = Field(..., min_length=1, max_length=255)
    sequence_id: str | None = Field(None, max_length=36)
    step_id: str = Field("", max_length=255)
    atml_ref: str | None = Field(None, max_length=255)
    status: str = Field("draft", max_length=20)


class TestCaseResponse(TestCaseCreate):
    """A persisted test case."""

    model_config = {"from_attributes": True}

    id: str
    created_at: datetime
    updated_at: datetime


# ── Read APIs: paged lists, traceability, graph browse (frontend tasks 25/26) ─


class RequirementPage(BaseModel):
    """Paged TestRequirement list envelope (same {items,total} shape as fmea)."""

    items: list[TestRequirementResponse]
    total: int


class CaseResponse(TestCaseResponse):
    """A test case joined to its requirement for the traceability matrix.

    ``product_code`` / ``requirement_code`` are denormalized from the linked
    requirement so the matrix can filter/group without a second round-trip;
    they are ``None`` for an unlinked (orphan) case.
    """

    product_code: str | None = None
    requirement_code: str | None = None


class CasePage(BaseModel):
    """Paged TestCase list envelope."""

    items: list[CaseResponse]
    total: int


class TraceabilityCase(BaseModel):
    """One case row inside the traceability tree (case → DSL step mapping)."""

    model_config = {"from_attributes": True}

    id: str
    case_code: str
    title: str
    sequence_id: str | None = None
    step_id: str = ""
    atml_ref: str | None = None
    status: str = "draft"


class TraceabilityRequirement(BaseModel):
    """A requirement with its verifying cases (requirement → cases → steps)."""

    id: str
    requirement_code: str
    title: str
    source: RequirementSource
    cases: list[TraceabilityCase] = []


class TraceabilityTree(BaseModel):
    """Structured requirement→cases→DSL-step tree for the matrix view.

    ``unlinked_cases`` holds cases whose ``requirement_id`` is unset (ingested
    before their requirement) so traceability gaps stay visible rather than
    being silently dropped.
    """

    product_code: str | None = None
    requirements: list[TraceabilityRequirement] = []
    unlinked_cases: list[TraceabilityCase] = []


class GraphNode(BaseModel):
    """A graph node shaped for a visualization UI.

    ``label`` is the primary node label/type; ``type`` mirrors it for graph
    libraries that key on ``type``; ``properties`` carries the remaining
    node properties.
    """

    id: str
    label: str
    type: str
    name: str = ""
    properties: dict[str, object] = Field(default_factory=dict)


class GraphEdge(BaseModel):
    """A directed relationship shaped for a visualization UI."""

    source: str
    target: str
    type: str


class GraphBrowse(BaseModel):
    """{nodes, edges} payload for the knowledge-graph browse view (task 25)."""

    nodes: list[GraphNode] = []
    edges: list[GraphEdge] = []


# ── FMEA ────────────────────────────────────────────────────────────────────


class FMEABase(BaseModel):
    """Shared FMEA fields. Ratings are 1-10; ``rpn`` is never an input."""

    component_code: str = Field(..., min_length=1, max_length=200)
    function_name: str | None = Field(None, max_length=255)
    fault_code: str | None = Field(None, max_length=100)
    failure_mode: str = Field(..., min_length=1, max_length=500)
    effects: str | None = None
    cause: str | None = None
    severity: int = Field(..., ge=RATING_MIN, le=RATING_MAX)
    occurrence: int = Field(..., ge=RATING_MIN, le=RATING_MAX)
    detection: int = Field(..., ge=RATING_MIN, le=RATING_MAX)
    recommended_action: str | None = None


class FMEACreate(FMEABase):
    """Create payload for an FMEA entry (no rpn — derived server-side)."""


class FMEAUpdate(BaseModel):
    """Partial-update payload for an FMEA entry (PUT/PATCH).

    Every field is optional; supplied ratings are still constrained to 1-10.
    ``rpn`` is deliberately absent — it is always derived server-side; a
    client-supplied ``rpn`` is ignored (extra fields are dropped, as on
    create), never trusted.
    """

    component_code: str | None = Field(None, min_length=1, max_length=200)
    function_name: str | None = Field(None, max_length=255)
    fault_code: str | None = Field(None, max_length=100)
    failure_mode: str | None = Field(None, min_length=1, max_length=500)
    effects: str | None = None
    cause: str | None = None
    severity: int | None = Field(None, ge=RATING_MIN, le=RATING_MAX)
    occurrence: int | None = Field(None, ge=RATING_MIN, le=RATING_MAX)
    detection: int | None = Field(None, ge=RATING_MIN, le=RATING_MAX)
    recommended_action: str | None = None


class FMEAResponse(FMEABase):
    """A persisted FMEA entry with the computed RPN."""

    model_config = {"from_attributes": True}

    id: str
    rpn: int
    created_at: datetime
    updated_at: datetime


# ── Diagnosis ───────────────────────────────────────────────────────────────


class DiagnosisCreate(BaseModel):
    """Persist payload for a diagnosis (task 15). Feedback is captured later."""

    run_id: str | None = Field(None, max_length=36)
    session_id: str | None = Field(None, max_length=64)
    symptom: str = Field(..., min_length=1)
    conclusion: str | None = None
    context_summary: str | None = None
    llm_model: str | None = Field(None, max_length=100)


class DiagnosisFeedback(BaseModel):
    """Operator feedback on a diagnosis (confirm/reject + optional note)."""

    helpful: bool
    feedback_note: str | None = None


class DiagnosisResponse(DiagnosisCreate):
    """A persisted diagnosis, including any operator feedback."""

    id: str
    helpful: bool | None = None
    feedback_note: str | None = None
    created_at: datetime
    updated_at: datetime


# ── ATML TestDescription import (task 11) ───────────────────────────────────


class ATMLImportCounts(BaseModel):
    """Created/updated counters for one entity kind during an import."""

    created: int = 0
    updated: int = 0


class ATMLUnmappedCase(BaseModel):
    """A test case persisted without a DSL step link (traceability gap)."""

    case_code: str
    title: str
    reason: str


class ATMLImportSummary(BaseModel):
    """Result summary returned by POST /api/v1/atml/import-test-description."""

    product_code: str
    requirements: ATMLImportCounts
    cases: ATMLImportCounts
    unmapped: list[ATMLUnmappedCase] = []


# ── Knowledge extraction (task 12) ──────────────────────────────────────────


class KnowledgeExtractRequest(BaseModel):
    """Trigger deterministic extraction from structured server-side sources.

    Paths are server-local filesystem paths (DSL YAML plans and recordings
    JSONL); no file upload — python-multipart is not installed. ATML import
    uses its own raw-XML endpoint.
    """

    product_code: str = Field(..., min_length=1, max_length=100)
    dsl_paths: list[str] = Field(default_factory=list)
    recording_paths: list[str] = Field(default_factory=list)


class SourceExtractCounts(BaseModel):
    """Created/updated counters for one extraction source."""

    created: int = 0
    updated: int = 0


class KnowledgeExtractSummary(BaseModel):
    """Result summary returned by POST /api/v1/knowledge/extract."""

    product_code: str
    requirements: SourceExtractCounts
    cases: SourceExtractCounts
    recordings_read: int = 0
    results_written: int = 0
    recording_events_skipped: int = 0
    recordings_skipped: list[str] = []
    unmatched_steps: list[str] = []
    graph_status: str = "ok"  # "ok" | "degraded" (no graph backend configured)

File: frontend/src/api/knowledge.ts
/**
 * Knowledge-graph read API module (tasks 25 & 26).
 *
 * Typed client over the backend knowledge READ endpoints (mounted on the
 * existing knowledge router, JWT-protected):
 *
 * - `GET /knowledge/graph`         — { nodes, edges } for the graph-browse UI;
 *   the backend returns an honest 503 when the graph backend is absent/down.
 * - `GET /knowledge/traceability`  — requirement → cases → DSL-step tree.
 * - `GET /knowledge/requirements`  — paged TestRequirement list ({items,total}).
 * - `GET /knowledge/cases`         — paged TestCase list joined to requirement.
 *
 * All transport uses the shared axios instance (`@/api/interceptor`) with the
 * JWT interceptor + 401 refresh — no bare `axios.create` / raw fetch.
 *
 * Backend contracts: src/ate_cloud/api/v1/knowledge_reads.py,
 * src/ate_cloud/schemas/knowledge.py.
 */
import http from './interceptor'

const api = http

// ── Graph browse (task 25) ───────────────────────────────────────────────────

/** A graph node shaped for visualization (mirrors backend GraphNode). */
export interface GraphNode {
  id: string
  /** Primary node label/type. */
  label: string
  /** Mirrors `label` for graph libraries that key on `type`. */
  type: string
  /** Human-readable name (may be empty). */
  name: string
  /** Remaining node properties. */
  properties: Record<string, unknown>
}

/** A directed relationship (mirrors backend GraphEdge). */
export interface GraphEdge {
  source: string
  target: string
  type: string
}

/** `{ nodes, edges }` payload returned by GET /knowledge/graph. */
export interface GraphBrowse {
  nodes: GraphNode[]
  edges: GraphEdge[]
}

/** Query params for the graph browse endpoint. */
export interface GraphBrowseParams {
  limit?: number
  label?: string
}

/** Fetch the knowledge graph for browsing; rejects with 503 when unavailable. */
export async function fetchKnowledgeGraph(
  params: GraphBrowseParams = {},
): Promise<GraphBrowse> {
  const query: Record<string, string | number> = {}
  if (params.limit !== undefined) query.limit = params.limit
  if (params.label) query.label = params.label
  const response = await api.get<GraphBrowse>('/knowledge/graph', { params: query })
  return response.data
}

// ── Requirements / cases paged lists (task 26) ───────────────────────────────

/** A persisted test requirement (mirrors backend TestRequirementResponse). */
export interface TestRequirement {
  id: string
  product_code: string
  requirement_code: string
  title: string
  description: string | null
  source: 'dsl' | 'atml' | 'manual'
  atml_ref: string | null
  created_at: string
  updated_at: string
}

/** A test case joined to its requirement for the traceability matrix. */
export interface TestCase {
  id: string
  requirement_id: string | null
  case_code: string
  title: string
  sequence_id: string | null
  step_id: string
  atml_ref: string | null
  status: string
  created_at: string
  updated_at: string
  /** Denormalized from the linked requirement (null for an orphan case). */
  product_code: string | null
  /** Denormalized from the linked requirement (null for an orphan case). */
  requirement_code: string | null
}

/** Paged list envelope ({items,total}) shared by the requirements/cases lists. */
export interface Paged<T> {
  items: T[]
  total: number
}

/** Query params for GET /knowledge/requirements. */
export interface RequirementListParams {
  product_code?: string
  source?: string
  skip?: number
  limit?: number
}

/** Query params for GET /knowledge/cases. */
export interface CaseListParams {
  requirement_id?: string
  product_code?: string
  skip?: number
  limit?: number
}

/** Fetch a paged list of test requirements. */
export async function fetchRequirements(
  params: RequirementListParams = {},
): Promise<Paged<TestRequirement>> {
  const query: Record<string, string | number> = {}
  if (params.product_code) query.product_code = params.product_code
  if (params.source) query.source = params.source
  if (params.skip !== undefined) query.skip = params.skip
  if (params.limit !== undefined) query.limit = params.limit
  const response = await api.get<Paged<TestRequirement>>('/knowledge/requirements', {
    params: query,
  })
  return response.data
}

/** Fetch a paged list of test cases (joined to their requirement). */
export async function fetchCases(params: CaseListParams = {}): Promise<Paged<TestCase>> {
  const query: Record<string, string | number> = {}
  if (params.requirement_id) query.requirement_id = params.requirement_id
  if (params.product_code) query.product_code = params.product_code
  if (params.skip !== undefined) query.skip = params.skip
  if (params.limit !== undefined) query.limit = params.limit
  const response = await api.get<Paged<TestCase>>('/knowledge/cases', { params: query })
  return response.data
}

// ── Traceability tree (task 26) ──────────────────────────────────────────────

/** One case row inside the traceability tree (case → DSL step mapping). */
export interface TraceabilityCase {
  id: string
  case_code: string
  title: string
  sequence_id: string | null
  step_id: string
  atml_ref: string | null
  status: string
}

/** A requirement with its verifying cases (requirement → cases → steps). */
export interface TraceabilityRequirement {
  id: string
  requirement_code: string
  title: string
  source: 'dsl' | 'atml' | 'manual'
  cases: TraceabilityCase[]
}

/** requirement → cases → DSL-step tree returned by GET /knowledge/traceability. */
export interface TraceabilityTree {
  product_code: string | null
  requirements: TraceabilityRequirement[]
  /** Cases ingested without a linked requirement (traceability gaps). */
  unlinked_cases: TraceabilityCase[]
}

/**
 * Fetch the requirement → cases → DSL-step traceability tree.
 *
 * Cases without a requirement land in `unlinked_cases` so matrix gaps (cases
 * ingested before their requirement) stay visible.
 */
export async function fetchTraceability(
  productCode?: string,
): Promise<TraceabilityTree> {
  const params: Record<string, string> = {}
  if (productCode) params.product_code = productCode
  const response = await api.get<TraceabilityTree>('/knowledge/traceability', { params })
  return response.data
}
